docs/signing-tool-spec.md

```diff
@@ -0,0 +1,129 @@
+
+With several different code signing tools required, each with its own command line parameters and defaults. The process of code signing is error-prone and hard to get right. The signing tool addresses these challanges by unifying the interface into a single set of options.
+
+## Plan
+
+Create a modern signing tool to replace the existing tools. The tool will handle all of our first party signing formats, orchestrate signing files in the right order, and have extensibility to support multiple signing providers. As our customers use a variety of clouds and HSMs, the extensibility will enable us to meet customers needs wherever they store their certificates.
+
+While some of this could be done via an MSBuild task, a CLI tool is preferable to MSBuild tasks for a couple reasons:
+
+- **Performance:** During a build, many binary artifacts are created that need to be signed. A multi-targeted NuGet package may contain several `.dll` files. An application will likely contain more than one file that needs to be signed. It's much more efficient to pass them all to a signing tool where parallelism is possible than to sign during the inner-loop.
+- **Security:** Code signing is a sensitive operation that requires credentials/secrets. We do not want secrets to be in the MSBuild pipeline as that makes any logs contain those secrets. Ideally, a CI pipeline should contain a separate stage for code signing to ensure that credentials are never unintentionally exposed to a build stage.
+- **Platform:** Authenticode is currently limted to Windows. Thus, while it's possible to sign a NuGet or VSIX cross-platform, the DLL's inside can't be signed unless running on Windows. With the NuGet packages being developer-only artifacts--they're not shipped with the apps--it's critical that the DLL's inside are also signed. Builds for binaries may run on any platform, but as signing is a discreet step in most CI pipelines, it's possible to use a Windows agent for this task.
+
+### Roadmap
+
+The scope of the preview release will be limited to the existing funtionality currently in the service. The remaining functionality in this spec will be delivered for the initial release. The .NET Foundation has a dependency on this tool being delivered by the end of 2022.
+
+#### Preview
+
+**Goals**
+
+- Support for Authenticode, VSIX, NuGet, ClickOnce
+- Only run on Windows x64.
+- Support a single certificate for all files in the operation.
+
+**Non-Goals**
+
+- Strong Name signing won't be in v1; guidance is to use an snk not based on a cert. If easy, perhaps can revisit.
+- Support for Microsoft internal builds and signing services.
+- Containers, including Notary v2 support.
+- Extensibility. v1 will support different signing providers.
```

Comment:
It seems like there are (at least) two extensibility pivots: one is the signing provider (the thing that holds the certificate/HSM/secret) and one is the artifact type to be signed. I see the former is planned for extensibility, what about the latter? How would someone extend this tool to sign their own thing (e.g. other package manager artifacts)?

Reply:
The latter isn't currently planned for external extensibility. Given the structure of the code, potential nesting of containers and file types, an internal integration might be the best option. In that case, a pull request with the to add the artifact type would be the best option. The codebase already has an interface that it'd implement and a way to declare file extensions it supports.

Reply:
Makes sense. It could be worth calling that out in the spec.